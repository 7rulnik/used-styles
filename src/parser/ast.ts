import {CodeLocation} from "./ranges";

export interface StyleRule {
  prop: string;
  value: string;
}

export interface StyleSelector {
  selector: string;
  postfix: string;

  pieces: string[];
  media: string[];

  declaration: number;
}

export interface StyleBody {
  id: number;
  rules: Array<StyleRule>;
  start: CodeLocation;
  end: CodeLocation;
}

export type StyleBodies = Record<number, StyleBody>;

export interface AST {
  selectors: StyleSelector[];
  bodies: StyleBodies;
}