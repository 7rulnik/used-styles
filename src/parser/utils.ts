export const mapStyles = (styles: string) => (
  (
    styles
    // remove style body
      .replace(/({[^{}]+})/g, '$')
      .replace(/({[^{}]+})/g, '$')
      // match style name
      .match(/\.([^>~,$:{\[\s]+)?/g) || []
  )
  // clean style name
    .map(x => x.replace(/[\s,.>~$]+/, ''))
    .map(x => x.replace(/[.\s.:]+/, ''))
);