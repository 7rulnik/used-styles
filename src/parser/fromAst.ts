import {SingleStyleAst, StyleBody, StyleRule, StyleSelector} from "./ast";

let separator = process.env.NODE_ENV === 'production' ? '' : '\n';

export const escapeValue = (value: string, name: string) => {
  if (name == 'content') return value.split('\\').join('\\\\');
  return value;
};

const createDecl = (decl: StyleRule) => (
  decl.prop + ': ' + escapeValue(decl.value, decl.prop) + ';'
);

const declsToString = (rules: StyleRule[]) =>
  rules
    .map(decl => createDecl(decl))
    .join(separator);

const getMedia = ({media}: { media: string[] }) => {
  const prefix: string[] = [];
  const postfix: string[] = [];

  media.forEach(media => {
    prefix.push(`@media ${media} {`);
    postfix.push('}');
  });
  return [
    prefix.join(separator),
    postfix.join(separator),
  ]
};


const renderRule = (rule: StyleSelector, style: StyleBody) => (
  `${rule.selector} { ${declsToString(style.rules)} }`
);

const isMatching = (selector: string, rule: StyleSelector) => (
  rule.pieces.some(piece => piece === selector)
);

const findMatchingSelectors = (selector: string, selectors: StyleSelector[]): StyleSelector[] => (
  selectors.filter(rule => isMatching(selector, rule))
);

export const fromAst = (rules: string[], {selectors, bodies}: SingleStyleAst, filter?: (selector: string) => boolean) => {

  const blocks: StyleSelector[] = [];

  rules.forEach(rule => {
    blocks.push(
      ...findMatchingSelectors(rule, selectors)
        .filter(block => !filter || filter(block.selector))
    );
  });

  blocks.sort((ruleA, ruleB) => bodies[ruleA.declaration].id - bodies[ruleB.declaration].id);

  const result: string[] = [];

  let lastMedia = ['', ''];
  blocks.forEach((block, index) => {
    const media = getMedia(block);
    if (media[0] !== lastMedia[0]) {
      result.push(lastMedia[1]);
      lastMedia = media;
      result.push(lastMedia[0]);
    }
    if (index < blocks.length - 1 && block.declaration === blocks[index + 1].declaration) {
      result.push(`${block.selector},`);
    } else {
      result.push(renderRule(block, bodies[block.declaration]));
    }
  });
  result.push(lastMedia[1]);

  return result.join(separator);
};