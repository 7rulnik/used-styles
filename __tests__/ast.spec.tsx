import * as React from 'react';
import {buildAst} from "../src/parser/toAst";
import {fromAst} from "../src/parser/fromAst";

describe('test ast', () => {
  const CSS = `
      .a,
      .b .c {
        border:1px solid;
        margin: 6px 10px;
      }
      
      .d ~ .e:not(focused){ 
        display: block;
        position: relative;
        width: calc(100% - 10px);
      }
      
      @media only screen and (max-width: 600px) {
        .c { position: absolute; }
        .a { position: relative; }  
      }
    `;

  it('should map simple style', () => {
    const ast = buildAst(CSS);
    expect(ast).toMatchSnapshot();
  });

  it('should remap simple style', () => {
    const ast = buildAst(CSS);
    const css = fromAst(['d'], ast);
    expect(css).toMatchSnapshot();
  });

  it('should remap complex style', () => {
    const ast = buildAst(CSS);
    const css = fromAst(['a', 'c'], ast);
    console.log(css);
    expect(css).toMatchSnapshot();
  });
});
