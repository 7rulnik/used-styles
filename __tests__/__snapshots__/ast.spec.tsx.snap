// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ast should map simple style 1`] = `
Object {
  "bodies": Object {
    "1": Object {
      "end": Object {
        "column": 25,
        "file": 0,
        "line": 5,
      },
      "id": 1,
      "rules": Array [
        Object {
          "prop": "border",
          "value": "1px solid",
        },
        Object {
          "prop": "margin",
          "value": "6px 10px",
        },
      ],
      "start": Object {
        "column": 9,
        "file": 0,
        "line": 4,
      },
    },
    "2": Object {
      "end": Object {
        "column": 33,
        "file": 0,
        "line": 11,
      },
      "id": 2,
      "rules": Array [
        Object {
          "prop": "display",
          "value": "block",
        },
        Object {
          "prop": "position",
          "value": "relative",
        },
        Object {
          "prop": "width",
          "value": "calc(100% - 10px)",
        },
      ],
      "start": Object {
        "column": 9,
        "file": 0,
        "line": 9,
      },
    },
    "3": Object {
      "end": Object {
        "column": 32,
        "file": 0,
        "line": 15,
      },
      "id": 3,
      "rules": Array [
        Object {
          "prop": "position",
          "value": "absolute",
        },
      ],
      "start": Object {
        "column": 14,
        "file": 0,
        "line": 15,
      },
    },
    "4": Object {
      "end": Object {
        "column": 32,
        "file": 0,
        "line": 16,
      },
      "id": 4,
      "rules": Array [
        Object {
          "prop": "position",
          "value": "relative",
        },
      ],
      "start": Object {
        "column": 14,
        "file": 0,
        "line": 16,
      },
    },
  },
  "selectors": Array [
    Object {
      "declaration": 1,
      "media": Array [],
      "pieces": Array [
        "a",
      ],
      "postfix": "",
      "selector": ".a",
    },
    Object {
      "declaration": 1,
      "media": Array [],
      "pieces": Array [
        "b",
        "c",
      ],
      "postfix": ".c",
      "selector": ".b .c",
    },
    Object {
      "declaration": 2,
      "media": Array [],
      "pieces": Array [
        "d",
        "e",
      ],
      "postfix": "~ .e:not(focused)",
      "selector": ".d ~ .e:not(focused)",
    },
    Object {
      "declaration": 3,
      "media": Array [
        "only screen and (max-width: 600px)",
      ],
      "pieces": Array [
        "c",
      ],
      "postfix": "",
      "selector": ".c",
    },
    Object {
      "declaration": 4,
      "media": Array [
        "only screen and (max-width: 600px)",
      ],
      "pieces": Array [
        "a",
      ],
      "postfix": "",
      "selector": ".a",
    },
  ],
}
`;

exports[`test ast should remap complex style 1`] = `
".a,
.b .c { border: 1px solid;
margin: 6px 10px; }

@media only screen and (max-width: 600px) {
.c { position: absolute; }
.a { position: relative; }
}"
`;

exports[`test ast should remap simple style 1`] = `
".d ~ .e:not(focused) { display: block;
position: relative;
width: calc(100% - 10px); }
"
`;
